### DISCUSSION TITLE
Automation to add/sync speakers to current one playing

### DISCUSSION THREAD

**Discussion by MdeBoer3** [music-assistant/support#1609] [Feature requests and ideas]
### What version of Music Assistant has the issue?

MA 2.0.0b74 HA 2023.11.1

### The problem

Hi all,

First of all, big fan of MA. Thanks all for the huge effort.

The question/request I have:
Would it be possible to add/remove speakers to the current one already playing music, using automations?

I know it's possible to create groups in MA and also how to add/remove speakers manually.
The situation is: I would like to have one 'main' speakers always playing, and based on some automations add/remove others players. Without the music/playlist being interrupted and so keeping the current playlist playing. 

So, basically, how to automate this command:
<img width="276" alt="Schermafbeelding 2023-11-06 om 08 25 53" src="https://github.com/music-assistant/hass-music-assistant/assets/60851589/c2abdbcb-8c1f-4d98-bb1f-8acf2dcf6295">

### How to reproduce

N/A

### Additional information

Current version: 2.0.0b74 

### What version of Home Assistant Core are your running

2023.11.1

### What type of installation are you running?

Home Assistant OS

### On what type of hardware are you running?

Alternative

**Comment by marcelveldt** [comment]
The service call to do that will eventually be added to the HA integration, it's one of the leftover things to fix before we hit stable.
You can already do this by using MA groups. Just create a group with the speakers in it and MA will take of sync/unsync when a player is powered on/off. Just power on the "master" player within the group and then power on others during the day. It will give you exact the effect you want.

**Reply by marekpytel** [reply]
Hi Community, 
First of all, Dear MASS Developers, thank you for inventing, implementing and maintaining this project:) For myself, been looking for 2 years a solution to play my liked music playlist on a Google speaker/Chromecast flavor of player, at a single button click, the Music Assistant add-on is a huge thing:) 
I'm new here, however been participating in an issue report of randomly looping YT track on a Chromecast player.

As for the idea, which seems to be coherent to the one that started this thread, somewhere in future release, could you try to "emulate" a YT Music player or "delegate" YT playlist handling to the 'target player' (or random player from the G-group) when Google-flavour target player is selected, so that the Google ecosystem can 'take over' control on an already triggered playback? 
Why so? I think there will be a couple benefits from that way of handling it: 1. the playback will start almost immediately (from what I'm observing by playlist started from either one of YT apps or a Google Nest2 directly, via either touch-selecting tracks or a voice-command issued to G-Assistant, that is the case). 2. In a Google Home app you'll be able to join/remove further players/spekers to the currently playing music. 3. Using either of the controls out of Google Home (YT App, voice, GH App, touch screen on Nest2) you'll be able to navigate playlist.

My actual use case is the Home Assistant MASS add-on, multiple Google-flavor players, YT Premium as a main media source (playlists), I'm using MASS mainly to provide YT playlists as a media source available in HA Automations (for the "duplicate player" entities)  to be triggered by different HA events.

Do I make sense here?

PS I also think that passing YT playback to GH components would work around the issues with looped or not being able to start tracks.

**Reply by OzGav** [reply]
Actually I don't think I am following what you are suggesting. If I understand your PS I don't believe handing off playback to some other device is possible as MA is creating the stream that is played.

**Reply by barrelltitor** [reply]
The current implementation of groups means that a speaker can only be part of one active group at a time. So, if you want to have two people being able to control them and listen to music at the same time and "move the music" around different speakers this would not be possible without first removing the speaker from the group and adding it to the other group.

Let's say you want to get a coffee in the kitchen and your wife is putting her makeup on in the bathroom. You're both listening to different music/podcasts/whatnot. 

You got your coffee, so you head to the living room, so you turn on the Living Room Speaker and turn off the Kitchen Speaker. If your wife now wants to use the Kitchen Speaker she needs to remove it from your group and add it to hers

If you want to continue from the same timestamp on another speaker you need to manually set it on the new speaker, or add the group to the speaker, which is quite annoying to do versus just turning the speaker "off" and turning another one "on" in the group.

I don't know if this is what OP wants, but I believe a button to "activate" or "deactivate" a speaker in a group without fully removing it would work wonders, and allowing this to be accessed via API or the HA integration would probably achieve what OP wants.

**Reply by marcelveldt** [reply]
Well, we can't. This is a limitation of google cast groups. If you want more flexibility, I'd suggest Sonos or Airplay speakers or if you want to go the DIY route, players based on squeezelite or snapcast.

**Reply by barrelltitor** [reply]
@marcelveldt I wasn't suggesting this for chromecast, I am indeed using squeezelite on all players, and the above example was assuming a DIY route such as squeezelite, my bad for not mentioning that

Edit: for anyone facing the same issue, just remove the groups, then you can sync the squeezelite players directly to each other with a sync button that will appear, which basically achieves this

If you're using chromecast, most of those devices support bluetooth, so you can use squeezelite on something connected to the speaker.

**Comment by OzGav** [comment]
In relation to the original feature request by @MdeBoer3 I believe this has now been implemented as the HA integration supports the `media_player.join` and `media_player.unjoin` service calls.

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
